Test code 작성하기

0-Chan opened the pull request:
안녕하세요~~

먼저 강의 영상 보고나서 `List.jsx` 에 대한 테스트 코드를 작성 해보았습니다^^

수정할 것이 있을 것 같은데.. 코드 리뷰 부탁드리겠습니다!

감사합니다.

0-Chan committed:
Add List test code

0-Chan committed:
Fix the error from eslint

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,41 @@
+
+import List from './List';
+
+test('List', () => {
+  const tasks = [
+    { id: 1, title: '뭐라도 하기' },
+    { id: 2, title: '코드숨 과제' },
+  ];
+
+  const handleClick = jest.fn();
+
+  const { container, getAllByText, rerender } = render((
+    <List
+      tasks={tasks}
+      onClickDelete={handleClick}
+    />
+  ));
+
+  expect(container).toHaveTextContent('뭐라도 하기');
+  expect(container).toHaveTextContent('코드숨 과제');
+  expect(container).toHaveTextContent('완료');
+
+  expect(handleClick).not.toBeCalled();
+
+  const buttons = getAllByText('완료');
+  fireEvent.click(buttons[0]);
+  fireEvent.click(buttons[1]);
+
+  expect(handleClick).toBeCalledWith(1);
+  expect(handleClick).toBeCalledWith(2);
+
+  tasks.length = 0;
+  rerender((
+    <List
+      tasks={tasks}
+      onClickDelete={handleClick}
+    />
+  ));
+  expect(container).toHaveTextContent('할 일이 없어요!');
+});
```
- 지금 총 3가지 테스트를 분리하지 않고 1개의 테스트에 작성하신 것 같아요!
    - 테스트 1 : 할일이 있을때
    - 테스트 2 : 완료버튼 클릭할 때 
    - 테스트 3 : 할일이 없을 때
- describe-(context)-it 패턴을 사용해서 1개의 테스트에 1개의 테스트 목적이 되도록 여러개의 테스트 시나리오로 쪼개봅시다!
`describe` 는 테스트 대상을 서술하고, 
`context` 는 when, with, without 만을 사용해 주어진 상황(맥락)을 서술하며, 
`it` 은 예상되는 동작(=테스트 내용)을 서술합니다.

0-Chan replied:
아하..! 분리하여 테스트 코드를 짰을 때 훨씬 가독성이 높아지는걸 볼 수 있었습니다!

daadaadaah reviewed (changes requested):
월요일부터 PR 날려주셨네요! 아주 좋습니다👍

[BDD 스타일](https://ko.javascript.info/testing-mocha#ref-1068)을 사용해서 나머지 테스트 코드도 작성해봅시다!💪

참고
- https://marocchino.net/2016/12/04/about-rspec/

0-Chan committed:
Apply BDD test code style

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,67 @@
+    });
+  });
+
+  context('tasks가 비어있다면', () => {
```
context는 대비되는 상황이 있을 때 사용되는 것이므로, 대비되는 테스트 다음에 위치하는게 좋을 것 같아요!ㅎㅎ

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,67 @@
+    });
+  });
+
+  context('완료 버튼을 누른다면', () => {
+    it('각각의 버튼에 함수가 동작한다.', () => {
+      const tasks = [
+        { id: 1, title: '뭐라도 하기' },
+        { id: 2, title: '코드숨 과제' },
+      ];
+
+      const { getAllByText } = render((
+        <List
+          tasks={tasks}
+          onClickDelete={handleClick}
+        />
+      ));
+
+      expect(handleClick).not.toBeCalled();
+
+      const buttons = getAllByText('완료');
+      fireEvent.click(buttons[0]);
+      fireEvent.click(buttons[1]);
+
+      expect(handleClick).toBeCalledWith(1);
+      expect(handleClick).toBeCalledWith(2);
+    });
+  });
```
1. [context](https://www.betterspecs.org/#contexts)는 대비되는 상황에서 사용하는 것인데, 완료 버튼을 누른다는 상황과 대비되는 상황이 없습니다. 따라서, 그냥 it문으로만 표현해주셔도 될 거예요!

2. 또한, 테스트 상황이 task가 존재할 때 테스트 할 수 있는 상황이므로, 아래 처럼 8번줄에 있는 테스트 묶어주셔도 될 것 같아요!
```jsx
    context('with tasks', () => {
      //... 생략
  
      it('renders tasks', () => {
          //... 생략
      });
  
      it('renders “완료” button to delete a task', () => {
          //... 생략
      });
    });
```

0-Chan replied:
짚어 주신 부분을 고려하여 test code를 작성 해보았습니다.

아 그리고 레퍼런스를 참고하여 영어로 작성 해보았는데... 어색한 부분이 있진 않을지 걱정이 되네요.ㅎㅎ

daadaadaah reviewed (changes requested):
잘하셨어요!👏👏 List.test.jsx 좀만 다듬고 다른 컴포넌트도 테스트 코드를 작성해보세요~!

1. 테스트를 영어로 작성해보시는 건 어떠신가요? ㅎㅎ어려우시겠지만 패턴이 정해져 있어서 한번 도전해봐도 괜찮을 것 같아요!ㅎㅎ
2. 테스트 코드를 다 작성하신 후에 중복이 없는지 한번 체크해보세요!

0-Chan committed:
Change Korean test code to English

0-Chan committed:
Simplify test code <List/> component

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,60 @@
+          tasks={tasks}
+          onClickDelete={handleClick}
+        />
+      ));
```
이코드가 반복되고 있습니다! 어떻게 하는게 좋을까요? 🤔

0-Chan replied:
<s>위 처럼 반복되는 것들은 모두 `const component`를 선언 후 할당하였습니다.</s>
이후 필요한 부분에서 호출하여 중복을 없앴습니다!

0-Chan replied:
- 수정
import 된 jsx를 바로 사용하는 쪽으로 바꾸었습니다.

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,60 @@
+      { id: 2, title: '코드숨 과제' },
+    ];
+
+    it('should renders tasks', () => {
```
테스트를 설명할 때, sholud를 빼고 작성해봅시다!ㅎㅎ
참고 : https://www.betterspecs.org/#should

0-Chan replied:
헛.. 이런 참고 자료를 찾아 헤매었는데 정말정말 감사드립니다! 많은 도움이 되었어요.ㅎㅎ

0-Chan committed:
Add Input.jsx test code

0-Chan committed:
Refactor Item.test.jsx to BDD style

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,37 @@
+  const handleClick = jest.fn();
+  const handleChange = jest.fn();
+
+  context('when component called', () => {
```
[context](https://www.betterspecs.org/#contexts)는 대비되는 상황에서 사용하는 것인데, 완료 버튼을 누른다는 상황과 대비되는 상황이 없습니다. 따라서, 그냥 it문으로만 표현해주셔도 충분합니다!

0-Chan replied:
context 문을 쓸 때는 대비되는 상황인지 보고 사용하겠습니다.

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,37 @@
+  const handleChange = jest.fn();
+
+  context('when component called', () => {
+    it('should render', () => {
```
무엇을 render 하는지 기재해주시는게 좋을 것 같아요!

0-Chan replied:
어떤 tag를 render 하는지 기재하였습니다!

daadaadaah commented on src/Item.test.jsx:
```diff
@@ -2,27 +2,36 @@ import { render, fireEvent } from '@testing-library/react';
+          onClickDelete={handleClick}
+        />
+      ));
+      expect(handleClick).not.toBeCalled();
+      fireEvent.click(getByText('완료'));
+      expect(handleClick).toBeCalledWith(1);
```
구분을 위해 한줄씩 띄어주시면 좋을 것 같아요!ㅎㅎ

0-Chan replied:
앗 ESLint에서 미처 잡지 못한 부분까지 신경 써주셔서 감사해요...!
비슷한 부분은 모두 가독성을 위해 띄워 놓았습니다.

daadaadaah commented on src/Item.test.jsx:
```diff
@@ -2,27 +2,36 @@ import { render, fireEvent } from '@testing-library/react';
+          task={task}
+          onClickDelete={handleClick}
+        />
+      ));
+      expect(container).toHaveTextContent('뭐라도 하기');
```
구분을 위해 한줄씩 띄어주시면 좋을 것 같아요!ㅎㅎ

daadaadaah reviewed (changes requested):
테스트 코드 작성 어느정도 감 익히신 것 같아서 기분이 좋네용👍
남겨드린 코멘트 참고하시면서 계속 진행해주세요🚀

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,37 @@
+  context('when component called', () => {
+    it('should render', () => {
+      const { container } = render((
+        <Input
```
value라는 props가 빠졌어요!ㅎㅎ

0-Chan replied:
value 값에 뭘 넣어야할지 고민하다 뺐었는데.. 생각해보니 props가 빠짐으로써 오히려 헷갈리게 할 수 있을 것 같네요.

`const value` 에 아무 문자열이나 할당하여 넣었습니다.

0-Chan committed:
:adhesive_bandage: Add line break for readable code

0-Chan committed:
:adhesive_bandage: Remove 'should' when describing tests

0-Chan committed:
:adhesive_bandage: Add const to prevent repetition

Add const called 'component' and assigned <List/>.

0-Chan committed:
:adhesive_bandage: Add missing props to <Input/>

0-Chan committed:
:adhesive_bandage: Remove unnecessary context part

Input.test.jsx test code is enough with 'it' statement.
And also added const called 'component' to prevent repetition.

0-Chan committed:
:sparkles: Add Page.jsx test code template

0-Chan committed:
:adhesive_bandage: Change const component to imported one

0-Chan committed:
:sparkles: Add input change test

Get <input> tag by placeholder and triggers fireEvent.

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,26 @@
+    expect(container).toHaveTextContent('추가');
+  });
+
+  it('renders clickable "추가" button', () => {
```
`clickable` 생략해주셔도 괜찮을 것 같아요!
`입력가능한 입력창을 렌더한다` 라는 느낌이라서, button이 주는 의미와 중복되는 느낌이거든요.
대신 무엇을 추가하는 버튼인지 설명해주셔도 괜찮을 것 같아요!
다른 동료가 봤을 때, 추가 버튼이라고 하면 무엇을 추가하는지가 궁금할 것 같거든요!

0-Chan replied:
지금 생각해보니 동어 반복이었네요 ㅋㅋㅋ 간단한 쪽으로 수정하였습니다.

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,43 @@
+      fireEvent.click(buttons[1]);
+
+      expect(handleClick).toBeCalledWith(1);
+      expect(handleClick).toBeCalledWith(2);
```
32번째 줄에서 2번 호출된 걸 파악하므로, 31번째 줄은 생략해도 괜찮을 것 같아요!

0-Chan replied:
넵 `toBeCalledTimes` 메서드로 바꾸어서 생략 하였습니다!

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,43 @@
+  });
+
+  context('without tasks', () => {
+    it('renders "할 일이 없어요!" <p> tag', () => {
```
`renders "할 일이 없어요!"` 로만 표현해주셔도 충분히 테스트 의도가 드러나는 것 같아요!
테스트 설명도 변수명처럼 간단히 할 수 있으면 간단히 하는게 좋아요!ㅎㅎ

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,49 @@
+  ];
+
+  context('with tasks', () => {
+    it('renders <h1>, <label>, <button> tags', () => {
```
html 요소 들이 render 된다는 표현 보다 `render "추가" 버튼` 이런식으로 표현해주는게 더 좋을 것 같아요!

0-Chan replied:
너무 세세한 표현 이었던 것 같네요. 짧지만 의도가 잘 드러나게 바꾸었습니다.

daadaadaah reviewed (changes requested):
1. 테스트 대상 컴포넌트들의 props 모두 기재해주는게 좋을 것 같아요!ㅎㅎ다른 동료가 처음 해당 컴포넌트를 봤을 때, 테스트마다 props가 다른 것과 모두 기재해 준 것과 어떤 것이 더 컴포넌트를 파악하기 쉬운지 생각해보시면 될 것 같아요!
2.  Page 컴포넌트 테스트는 작성중이신 것 같아서 다음 PR 때 다시 볼게요~ㅎㅎ

0-Chan committed:
:adhesive_bandage: Fix each component to have every props

0-Chan committed:
:adhesive_bandage: Fix 'it' statement for better expression

0-Chan committed:
:adhesive_bandage: Simplyfy multiple buttons click test

0-Chan committed:
:sparkles: Add left test codes for Page.jsx

0-Chan committed:
:sparkles: Add App.jsx test code

Also added "추가" test code for Page.test.jsx which was missing.

0-Chan commented:
> 1. 테스트 대상 컴포넌트들의 props 모두 기재해주는게 좋을 것 같아요!ㅎㅎ다른 동료가 처음 해당 컴포넌트를 봤을 때, 테스트마다 props가 다른 것과 모두 기재해 준 것과 어떤 것이 더 컴포넌트를 파악하기 쉬운지 생각해보시면 될 것 같아요!
> 2. Page 컴포넌트 테스트는 작성중이신 것 같아서 다음 PR 때 다시 볼게요~ㅎㅎ

마저 남은 Page와 App 컴포넌트 테스트를 모두 작성 했습니다.:tada:

테스트 코드를 짜며 들었던 의문이 2가지 있었는데요..

1. **<Page/> 에서 state의 변화를 테스트 하고 싶을 때**
예를 들면 Page에서 추가, 완료 버튼이 눌리느냐 뿐만 아니라, 실질적인 동작을 테스트 하고 싶은데요.<br>
현재 구조가 App 에서 state를 관리하고 이를 Page가 props로 받아서 처리하고 있는 상황에서 `jest.fn()` 으로 mock을 만들어서는 동작 테스트가 안되더라구요.<br>
그래서 문서에서 `jest.spyOn()`도 찾아보고 했는데.. 결국에는 안되는 것으로 보았거든요. 이런 방식의 테스트는 원래 안하는게 맞는건가요? 
<br><br>
2. **각각의 컴포넌트 마다의 테스트 코드 중복**
이건 그냥 궁금해서 여쭤보는건데요. Item도 List도 심지어 Page까지 "완료 버튼이 눌리는지" 를 테스트 하면서 <ins>이게 Item나 List 에서  이미 작동하는걸 확인했는데.. 굳이 다 확인 해야하나?</ins> 라는 의문이 들었는데 이런건 중복이 아닌건가요?

0-Chan committed:
:sparkles: Add input.value to be clear when button clicks.

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,57 @@
+    expect(input.value).toBe('Changed');
+  });
+
+  it('clicks "추가" button to display task', () => {
```
`display task`가 아니라 `add task`로 표현해주셔야 할 것 같아요!ㅎㅎ

0-Chan replied:
모호한 표현은 모두 수정 했습니다.ㅎㅎ

daadaadaah commented on src/Item.test.jsx:
```diff
@@ -2,27 +2,41 @@ import { render, fireEvent } from '@testing-library/react';
-  fireEvent.click(getByText('완료'));
-
-  expect(handleClick).toBeCalledWith(1);
+  context('with task', () => {
```
context는 대비대는 상황에서 사용하는게 적절해요! `with/without task` 상황이 아니므로 context 생략해주시고 it문으로만 테스트를 표현해주셔도 될 것 같아요!ㅎㅎ

0-Chan replied:
아하,.. 그렇네요 이게 참 context가 아직 와닿질 않은 것 같습니다.ㅜㅜ 앞으로 좀 더 연습 해볼께요.

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,56 @@
+    ];
+
+    it('renders tasks', () => {
+      const { container } = render(
+        <List
+          tasks={tasks}
+          onClickDelete={handleClick}
+        />,
+      );
```
테스트 코드를 다 작성하신 후 `리팩터링할 게 없나?` 라고 생각해보시는게 좋아요!
현재 이 코드가 반복되는데, 어떻게 하는게 좋을까요? 🤔 (다른 테스트 코드도 함께 고민해보세요!)
```jsx
render(
    <List
      tasks={tasks}
      onClickDelete={handleClick}
    />,
 );
```

0-Chan replied:
앞에도 짚어 주셨었는데 ㅜㅜㅜ 제가 정신이 없었네요. 

이렇게 반복되는 코드는 모두 변수에 넣어 처리 했습니다!

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,113 @@
+
+  it('renders <Input/> field', () => {
+    const { container } = render(
+      <Page
+        onChangeTitle={handleChange}
+        onClickAddTask={handleClickAdd}
+        tasks={tasks}
+        onClickDeleteTask={handleClickDelete}
+      />,
```
사소한 거지만, 데이터를 위쪽으로, 함수를 아래로 이렇게 정렬 맞춰주시는게 가독성이 더 좋을 것 같아요!ㅎㅎ
```jsx
      <Page
        tasks={tasks}
        onChangeTitle={handleChange}
        onClickAddTask={handleClickAdd}
        onClickDeleteTask={handleClickDelete}
      />,
```

0-Chan replied:
사소한 것 까지 알려주셔서 너무너무 감사합니다! 반영 했습니다.

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,113 @@
+  const handleClickAdd = jest.fn();
+  const handleClickDelete = jest.fn();
+
+  const tasks = [
+    { id: 1, title: '뭐라도 하기' },
+    { id: 2, title: '코드숨 과제' },
+  ];
```
지금 tasks 라는 테이터가 여러 테스트에서 사용되고 있어요!
이렇게 여러 곳에서 사용되는 데스트용 데이터를 보통 `fixture`라고 이야기합니다. 
이러한 데이터를 `fixtures` 폴더나 파일 하나 만들어서 따로 관리해주는 편입니다!
이렇게 했을 때 어떤 장점이 있을지에 대해 생각해보세요!

0-Chan replied:
`__fixtures__` 디렉토리를 하나 만들어서 적용 해보았습니다.

다음과 같은 장점을 얻은 것 같습니다.ㅎㅎ
- 테스트 코드의 길이가 줄어들어 가독성이 좋아집니다.
- 테스트 코드 재사용성이 올라가서 코드 품질이 좋아집니다.

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,56 @@
+        />,
+      );
+
+      expect(container).toHaveTextContent('뭐라도 하기');
+      expect(container).toHaveTextContent('코드숨 과제');
```
이렇게 쓰셔도 되고, tasks와 forEach를 이용해서 반복문 돌려도 괜찮아요!ㅎㅎ
저는 하나하나 써주기 귀찮아서 forEach 이용합니다.

0-Chan replied:
제가 바닐라 JS만 다뤘어서.. ES5, ES6 문법에 익숙치가 않았는데, 알려주신 방법으로 중복을 제거 했습니다. 감사해용~~

daadaadaah replied:
저도 코드숨 할 때, ES5 이후 문법 익숙하지 않았어요!
제가 읽고 괜찮았다고 생각한 책 하나 추천드립니다!
[코딩의 기술](https://book.interpark.com/product/BookDisplay.do?_method=detail&sc.prdNo=324153812&gclid=Cj0KCQjwr4eYBhDrARIsANPywCjv0UUT_zRT2DU5lh2mGLhMUrr4ykKS0KCaRc23zwXSFqYZvuKPDFIaAlGcEALw_wcB)

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,113 @@
+    { id: 2, title: '코드숨 과제' },
+  ];
+
+  it('renders <Input/> field', () => {
```
이 테스트 65째 줄 테스트랑 통합해주시는게 어떠신가요? 결국 Page 컴포넌트가 Input 컴포넌트랑 List 컴포넌트 통합해준 컴포넌트이므로, 그거에 맞게 테스트도 작성해주시는게 좋을 것 같아요!ㅎㅎ

0-Chan replied:
그렇네요.. 굳이 나눠서 볼 필요가 없었습니다.

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,113 @@
+    expect(container).toHaveTextContent('추가');
+  });
+
+  it('changes input value', () => {
```
이 테스트도 with task 상황에서 할 수 있는 행위이니, with tasks 안에 같이 넣어주시는게 테스트 파악하기 더 수월할 것 같아요!

0-Chan replied:
이 부분이 좀 헷갈리는데요 ㅜㅜ input을 수정하는 행위는 with task랑 상관 없이 이루어 지는게 아닐까요?

daadaadaah replied:
오잉? 제가 테스트를 "완료" 버튼 클릭 상황으로 착각하고, 잘못 드린 피드백이예요!😅
O-Chan님이 셍각하신게 맞아요! 
피드백에 대해 무비판적으로 받아드리시지 않고, 한번 생각해보신 거 너무 잘하셨어요!👏

daadaadaah reviewed (changes requested):
> Page 에서 state의 변화를 테스트 하고 싶을 때
> 예를 들면 Page에서 버튼이 눌리느냐 뿐만 아니라, 실질적인 동작을 테스트 하고 싶은데요.
>
> 현재 구조가 App 에서 state를 관리하고 이를 Page가 props로 받아서 처리하고 있는 상황에서 jest.fn() 으로 mock을 만들어서는 동작 테스트가 안되더라구요.
>
> 그래서 문서에서 jest.spyOn()도 찾아보고 했는데.. 결국에는 안되는 것으로 보았거든요. 이런 방식의 유닛 테스트는 원래 안하는게 맞는건가요? (혹시 e2e 테스트에서 담당하는 부분일까요?)

- 예를 들어, App 컴포넌트 처럼 완료 버튼을 눌렀을 때, 단순 handleClick 호출 대신 "할 일 없어요"라는 문구가 보인다 라는 테스트를 하고 싶었다는 이야기로 이해하면 될까요?ㅎㅎ

0-Chan committed:
:adhesive_bandage: Fix ambigous test case description

0-Chan committed:
:adhesive_bandage: Fix context to use it in proper way

0-Chan committed:
:adhesive_bandage: Add component render function to prevent duplicate

0-Chan committed:
:adhesive_bandage: Correct value attr to come first

0-Chan committed:
:sparkles: Add fixture for tasks

0-Chan committed:
:adhesive_bandage: Prevent repetition by using forEach

0-Chan committed:
:white_check_mark: Fix linter warnings for fixtures

0-Chan commented:
> > Page 에서 state의 변화를 테스트 하고 싶을 때
> > 예를 들면 Page에서 버튼이 눌리느냐 뿐만 아니라, 실질적인 동작을 테스트 하고 싶은데요.
> > 현재 구조가 App 에서 state를 관리하고 이를 Page가 props로 받아서 처리하고 있는 상황에서 jest.fn() 으로 mock을 만들어서는 동작 테스트가 안되더라구요.
> > 그래서 문서에서 jest.spyOn()도 찾아보고 했는데.. 결국에는 안되는 것으로 보았거든요. 이런 방식의 유닛 테스트는 원래 안하는게 맞는건가요? (혹시 e2e 테스트에서 담당하는 부분일까요?)
> 
> * 예를 들어, App 컴포넌트 처럼 완료 버튼을 눌렀을 때, 단순 handleClick 호출 대신 "할 일 없어요"라는 문구가 보인다 라는 테스트를 하고 싶었다는 이야기로 이해하면 될까요?ㅎㅎ

헉 정확합니다..! Page 컴포넌트가 잘 작동하는지 테스트하고 싶으면 "할 일 없어요"를 표시 해주는 것 까지 보여줘야 하는게 아닐까 싶어서요.

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,59 @@
+import App from './App';
+
+describe('<App/>', () => {
+  const appComponent = () => render(<App />);
```
함수 네이밍은 함수가 무슨 일을 하는지에 대해 명시해주는게 좋아요!
지금 상황에서 App 컴포넌트를 렌더하고 있으므로, `renderApp()` 라는 좀더 적절할 것 같아요! (다른 컴포넌트도 마찬가지구요!ㅎㅎ)

daadaadaah commented on src/Item.test.jsx:
```diff
@@ -1,28 +1,33 @@
-  const { container, getByText } = render((
+  const appComponent = (task) => render(
     <Item
+      key={task.id}
```
Item 컴포넌트 내에 선언된 Props 들만 기재해주셔도 될 것 같아요! 따라서, key는 생략해주셔도 됩니다!

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,71 @@
+      expect(handleClickDelete).not.toBeCalled();
+
+      const buttons = getAllByText('완료');
+      buttons.forEach((button) => fireEvent.click(button));
```
오! 다른 부분에도 적용하는 센스!👍

daadaadaah commented on src/__fixtures__/tasks.js:
```diff
@@ -0,0 +1,9 @@
+export const fixtureTasks = [
```
그냥 `tasks`로 표현해주셔도 괜찮아요!
어떤 느낌이냐면, 컴포넌트 네이밍 할 때, XXComponent를 다 기재해준 느낌이예요!ㅎㅎ

daadaadaah reviewed (changes requested):
>Page 컴포넌트가 잘 작동하는지 테스트하고 싶으면 "할 일 없어요"를 표시 해주는 것 까지 보여줘야 하는게 아닐까 싶어서요.

0-Chan 님 께서 파악하신 대로, App 컴포넌트에서 상태를 관리하고 있기 때문에, 행위로 인한 상태 변화는 App 컴포넌트 또는 E2E 테스트에서 합니다. Page 컴포넌트는 행위로 인한 상태변화가 아닌 호출 여부만 알 수 있으므로, 행위로 인한 상태변화를 테스트 할 수 없어요! 해당 내용은 4주차 과제 내용이기도 합니다!ㅎㅎ다양한 시도를 해보셔서 다음주차 내용을 미리 알고 가시네용👍👍
